src/coreclr/gc/gc.cpp

```diff
@@ -16162,13 +16169,24 @@ BOOL gc_heap::a_fit_segment_end_p (int gen_number,
                                  (end - allocated),
                                  gen_number, align_const);
 
-        if (grow_heap_segment (seg, (allocated + limit)))
+        if (grow_heap_segment (seg, (allocated + limit), &hard_limit_short_seg_end_p))
         {
             goto found_fit;
         }
         else
         {
-            *commit_failed_p = TRUE;
+#ifndef USE_REGIONS
+            if (!hard_limit_short_seg_end_p)
+            {
+                dprintf (2, ("can't grow segment, doing a full gc"));
+                *commit_failed_p = TRUE;
+            }
+            else
+#endif // !USE_REGIONS
+            {
+                *commit_failed_p = TRUE;
+                assert (heap_hard_limit);
```

Comment:
This is not quite right.

1. In the segment's case - we were **NOT** setting `*commit_failed_p` if we ran out of hard limit, now we will. That change is unintended.
2. In the region's case, we will be asserting `heap_hard_limit` even when we are simply running out of memory to commit.

Reply:
(1) is fixed, but (2) is still there.